### ISSUE TITLE
REST API to create output dataset lineage

### ISSUE THREAD

**Issue by krehermann** [MarquezProject/marquez#1516]
What is the supported mechanism to use the REST API to create lineage like input -> job -> output?

I've done a fair bit of hacking to get a lineage graph with inputs and output using the REST api but I can't seem to make it work. I am referencing the REST api docs from https://marquezproject.github.io/marquez/openapi.html

There are a few things that confuse me:
When I submit a PUT to create a job, the returned body does not contain the outputs that I submitted

```
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{
  "type": "BATCH",
  "inputs": [
    {
      "namespace": "test",
      "name": "people.names"
    }
  ],
  "outputs": [
    {
      "namespace": "test",
      "name": "people.addresses"
    }],
  "location": "http://notaplace",

  "description": "address mapper"
}' \
 'http://localhost:5000/api/v1/namespaces/test/jobs/address-lookup'
 
```
 
 returns
 
```
 {
  "id": {
    "namespace": "test",
    "name": "address-lookup"
  },
  "type": "BATCH",
  "name": "address-lookup",
  "createdAt": "2021-07-13T22:40:46.559895Z",
  "updatedAt": "2021-07-13T22:40:46.559895Z",
  "namespace": "test",
  "inputs": [
    {
      "namespace": "test",
      "name": "people.names"
    }
  ],
  "outputs": [],
  "location": "http://notaplace",
  "context": {},
  "description": "address mapper",
  "latestRun": null,
  "facets": {}
}
```

So, I looked at the database and I see the there is no column in the `jobs` table to represent outputs. Instead, it seems like all of this is the job_version cross reference tables. 

Now, I don't see anyway to create a job version via the REST API. Indeed, there are no job versions for my `address-lookup` test

```
 curl -i -X GET \
   -H "Content-Type:application/json" \
 'http://localhost:5000/api/v1/namespaces/test/jobs/address-lookup/versions'
> > HTTP/1.1 200 OK
Date: Tue, 13 Jul 2021 23:04:39 GMT
Content-Type: application/json
Vary: Accept-Encoding
Content-Length: 15

{"versions":[]}
```

I confirmed this by looking the database, too.

I created a run, but that had no effect on the job version table, as expected. 

What is the right way to do this?

Below is an embedded script the reproduces the entire story that I just wrote.

```

#/bin/bash

set -ex

#create namespace
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{"name":"test","ownerName":"coinbird","description":"test namespace!"}' \
'http://localhost:5000/api/v1/namespaces/test'

#create source
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{
  "type": "TestSource",
  "connectionUrl": "testing://hello",
  "description": ""
}' \
    'http://localhost:5000/api/v1/sources/sourcesrus'

#create input dataset
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{
  "type": "DB_TABLE",
  "physicalName": "mock.names",
  "sourceName": "sourcesrus",
  "fields": [
    {
      "name": "first",
      "type": "STRING"
    },
    {
      "name": "last",
      "type": "STRING"
    }
  ],
  "description": "whats in a name"
}' \
    'http://localhost:5000/api/v1/namespaces/test/datasets/people.names'

#create output dataset
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{
  "type": "DB_TABLE",
  "physicalName": "mock.address",
  "sourceName": "sourcesrus",
  "fields": [
    {
      "name": "city",
      "type": "STRING"
    },
    {
      "name": "state",
      "type": "STRING"
    }
  ],
  "description": "where its at"
}' \
    'http://localhost:5000/api/v1/namespaces/test/datasets/people.addresses'


#create job with input and output
curl -i -X PUT \
   -H "Content-Type:application/json" \
   -d \
'{
  "type": "BATCH",
  "inputs": [
    {
      "namespace": "test",
      "name": "people.names"
    }
  ],
  "outputs": [
    {
      "namespace": "test",
      "name": "people.addresses"
    }],
  "location": "http://notaplace",

  "description": "address mapper"
}' \
 'http://localhost:5000/api/v1/namespaces/test/jobs/address-lookup'

#create a run of the job
curl -i -X POST \
   -H "Content-Type:application/json" \
   -d \
'{
  "args": {
    "fuzzy": "bunny"
  }
}' \
    'http://localhost:5000/api/v1/namespaces/test/jobs/address-lookup/runs'

#see that there are no versions
curl -i -X GET \
   -H "Content-Type:application/json" \
 'http://localhost:5000/api/v1/namespaces/test/jobs/address-lookup/versions'
```

**Comment by krehermann** [comment]
I figured out that if I POST a run complete, then I get a job version, but the outputs are still empty, and the UI does not show any output lineage

```
curl -i -X POST \
   -H "Content-Type:application/json" \
    'http://localhost:5000/api/v1/jobs/runs/df143c47-9b37-495b-a7a5-f82af2a15e11/complete'
```

```

{
"id": "df143c47-9b37-495b-a7a5-f82af2a15e11",
"createdAt": "2021-07-13T23:02:45.579035Z",
"updatedAt": "2021-07-14T00:57:28.497354Z",
"nominalStartTime": null,
"nominalEndTime": null,
"state": "COMPLETED",
"startedAt": null,
"endedAt": "2021-07-14T00:57:28.497354Z",
"durationMs": null,
"args":{
"fuzzy": "bunny"
},
"jobVersion":{
"namespace": "test",
"name": "address-lookup",
"version": "9b9dfdab-e6ca-35d6-86ee-80ebbec2f71e"
},
"inputVersions":[
{
"namespace": "test",
"name": "people.names",
"version": "369aeadc-d7b7-397c-9cf7-2cecc835df56"
}
],
"outputVersions":[
],
"context":{},
"facets":{}
```

**Comment by wslulciuc** [comment]
@krehermann: You're absolutely! I used the `curl` calls outlined in your issue, and was able to reproduce the issue. I've added this issue to be fixed in our `0.17.0`, but we'll most likely have a fix out sooner.

**Comment by krehermann** [comment]
Fantastic, thank you. @wslulciuc can you confirm that this method **does not** produce a lineage graph in the UI? That is, I believe you have confirmed the `curl` request/responses, but I also want to know that you don't get output lineage in the UI. 

I have a attached a screenshot of what I see (there are a few more runs now, after more attempts to make something work)
![marquez-no-outputs](https://user-images.githubusercontent.com/16602512/125720768-68aedde2-68be-47bf-b1f6-efaa25678a97.png)

**Comment by wslulciuc** [comment]
@krehermann Yep, I was able to re-produce your lineage graph. So, after looking over the Marquez API spec, I realized that the lineage endpoint to accept [OpenLineage](https://openlineage.io) events was not documented, only the API to query the lineage graph. The main advantage of OpenLineage is that you only have to make a single API call to begin seeing a lineage graph for you job using an open standard. I've opened an issue to add the endpoint https://github.com/MarquezProject/marquez/issues/1526


Mind walkthrough the [getting started guide](https://openlineage.io/getting-started) for OpenLineage using Marquez as the HTTP backend? We'll still want to fix the bug outlined in this issue, but this will unblock you :)
